File: python-service/app/routes/address.py
"""Route for single address validation endpoint.

Exposes POST /validate/address which accepts a StructuredAddress,
validates required fields, and delegates to AddressValidator.
"""

from typing import List

from fastapi import APIRouter, Depends, HTTPException

from app.adapters.usps_adapter import USPSAdapterInterface, get_usps_adapter
from app.models.address import StructuredAddress
from app.models.validation import ValidationResult
from app.services.address_validator import AddressValidator

router = APIRouter()

# Required fields that must be non-empty for a valid address request
_REQUIRED_FIELDS = ["street_line_1", "city", "state", "zipcode"]


def _get_address_validator(
    adapter: USPSAdapterInterface = Depends(get_usps_adapter),
) -> AddressValidator:
    """Dependency that provides an AddressValidator wired with the USPS adapter."""
    return AddressValidator(usps_adapter=adapter)


def _get_missing_fields(address: StructuredAddress) -> List[str]:
    """Return list of required fields that are empty strings."""
    missing = []
    for field in _REQUIRED_FIELDS:
        value = getattr(address, field)
        if not value or value.strip() == "":
            missing.append(field)
    return missing


@router.post("/validate/address", response_model=ValidationResult)
async def validate_address(
    address: StructuredAddress,
    validator: AddressValidator = Depends(_get_address_validator),
) -> ValidationResult:
    """Validate a single structured address via USPS.

    Accepts a StructuredAddress in the request body. If any required fields
    (street_line_1, city, state, zipcode) are missing or empty, returns a
    400 error with the MISSING_FIELDS error code listing the missing fields.

    On success, delegates to AddressValidator.validate() and returns the
    ValidationResult.
    """
    missing = _get_missing_fields(address)
    if missing:
        raise HTTPException(
            status_code=400,
            detail={
                "error_code": "MISSING_FIELDS",
                "message": f"Missing required fields: {', '.join(missing)}",
                "fields": missing,
            },
        )

    return await validator.validate(address)

File: express-service/src/routes/address.ts
import { Router, Request, Response } from 'express';
import { StructuredAddress } from '../models/address';
import { AddressValidator } from '../services/addressValidator';
import { getCachedValidation, cacheValidation, logValidationRequest } from '../services/dynamoCache';

const REQUIRED_FIELDS: (keyof StructuredAddress)[] = [
  'street_line_1',
  'city',
  'state',
  'zipcode',
];

/**
 * Returns a list of required fields that are missing or empty in the request body.
 */
function getMissingFields(body: Record<string, unknown>): string[] {
  const missing: string[] = [];
  for (const field of REQUIRED_FIELDS) {
    const value = body[field];
    if (!value || (typeof value === 'string' && value.trim() === '')) {
      missing.push(field);
    }
  }
  return missing;
}

/**
 * Creates the address validation router.
 * POST /validate/address - Validate a single structured address via USPS.
 */
export function createAddressRouter(addressValidator: AddressValidator): Router {
  const router = Router();

  router.post('/validate/address', async (req: Request, res: Response) => {
    const body = req.body ?? {};

    // Check for missing required fields
    const missing = getMissingFields(body);
    if (missing.length > 0) {
      return res.status(400).json({
        detail: {
          error_code: 'MISSING_FIELDS',
          message: `Missing required fields: ${missing.join(', ')}`,
          fields: missing,
        },
      });
    }

    // Build the StructuredAddress from the request body
    const address: StructuredAddress = {
      street_line_1: body.street_line_1 as string,
      city: body.city as string,
      state: body.state as string,
      zipcode: body.zipcode as string,
      ...(body.street_line_2 ? { street_line_2: body.street_line_2 as string } : {}),
    };

    try {
      // Check DynamoDB cache first
      const cached = await getCachedValidation(address);
      if (cached) {
        // Log the cache hit
        logValidationRequest('/api/v1/validate/address', address, cached, 200);
        return res.status(200).json(cached);
      }

      // Cache miss — call USPS adapter
      const result = await addressValidator.validate(address);

      // Store in DynamoDB cache for future lookups
      cacheValidation(address, result);

      // Log the request
      logValidationRequest('/api/v1/validate/address', address, result, 200);

      return res.status(200).json(result);
    } catch (error) {
      return res.status(500).json({
        detail: {
          error_code: 'INTERNAL_ERROR',
          message: 'An unexpected error occurred during address validation',
        },
      });
    }
  });

  return router;
}
